### PR TITLE
OCPNODE-2387: SCC: add UserNamespaceLevel field

### PR DESCRIPTION
**Pull request by haircommander**
to allow an SCC to toggle whether a pod is forced to be confined by a user namespace

### CODE REVIEW

File: security/v1/types.go
Comment (JoelSpeed) [inline security/v1/types.go:99]: I think you missed my point about not using bools 😂 

We avoid bools because they are not expressive and do not allow future expansion. What you've done here is gamed the system to force a bool into an enum.

The idea would be that you come up with a descriptive field name, and some enum values that make sense, such that if we needed additional states in the future, we could add those.

It would be more like
```
hostUsers: Allowed | Disallowed | AllowedButOnlyIfXYZ | DisallowedExceptForWhenFoo
```

Comment (haircommander) [reply security/v1/types.go:99]: I thought the UX would be better if it could be expressed the same as the other AllowHost* fields in YAML. We still get the same level of expressiveness as we can in the future add AllowedButOnlyIfXYZ, but we maintain compatibility with existing fields. Are you against this approach??

Comment (JoelSpeed) [reply security/v1/types.go:99]: We try not to repeat sins of the past when we know better, and even with the enum, using true and false would break compatibility with existing types as `true != "true"` when you're parsing JSON

Comment (haircommander) [reply security/v1/types.go]: Okay I've switched to Allowed/Disallowed. Personally, I don't see a difference, but I'll defer to your wisdom here :)

File: security/v1/types.go
Comment (JoelSpeed) [inline security/v1/types.go:102]: I would make this into
```suggestion
	// hostUsers determines if the policy allows host users in containers.
	// Valid values are "Allowed", "Disallowed" and omitted.
	// When omitted, this means no opinion and the platform is left to choose a reasonable default.
	// The current default is "Allowed".
	// +openshift:enable:FeatureGate=UserNamespacesSupport
	// +kubebuilder:validation:Enum="Allowed";"Disallowed";""
	// +kubebuilder:validation:Optional
	HostUsers HostUsersStrategyType `json:"hostUsers" protobuf:"varint,26,opt,name=allowHostUsers"`
```

File: security/v1/types.go
Comment (JoelSpeed) [inline security/v1/types.go:291]: These will need to be updated

**Review by JoelSpeed — COMMENTED** [review]
I've lost the thread now, but for clarity on why we are switching to an enum and not using `"true"` and `"false"` as enum values.
* Bools cannot be expanded in the future, so we require enums as they can be expanded
* The string values `true` and `false`, in JSON would be confusing as the json marshaller would interpret them as bools, and fail to marshal into a string type field. However the strings `"true"` and `"false"` would marshal correctly.
* `true` and `false` are not PascalCase, which is the expected casing for enum values in APIs
* We prefer to not perpetuate errors in API design we have made in the past, and rather move forward with the things we know now

**Comment by JoelSpeed** [comment]
@ibihim @wallylewis are you both ok with this addition to SCC, I believe auth team would own this right?

File: security/v1/types.go
Comment (haircommander) [reply security/v1/types.go]: done!

File: security/v1/types.go
Comment (haircommander) [reply security/v1/types.go]: done!

File: security/v1/types.go
Comment (haircommander) [reply security/v1/types.go]: @rphillips WDYT about this? your comment contradicts...

Comment (haircommander) [reply security/v1/types.go]: https://github.com/openshift/enhancements/pull/1649#discussion_r1678338441

File: security/v1/types.go
Comment (JoelSpeed) [inline security/v1/types.go:99]:
```suggestion
	// When "RequirePodLevel" is set, a pod author must set `hostUsers` to `false`.
```

Comment (haircommander) [reply security/v1/types.go]: updated, thanks!

**Comment by JoelSpeed** [comment]
/test minor-e2e-upgrade-minor

Odd that the CLI image is missing suddenly, I don't _think_ that's a configuration problem

**Comment by haircommander** [comment]
/test /test minor-e2e-upgrade-minor 

has magic fixed it :thinking:

**Review by JoelSpeed — COMMENTED** [review]
A few marker things to fix and a comment, but otherwise I'm happy with this, seems the EP is going in the right direction as well, did you want @deads2k to scan too?

**Comment by haircommander** [comment]
updated, thanks @JoelSpeed ! Any reviews I can gather work for me, @deads2k do you have any additional notes?

**Comment by JoelSpeed** [comment]
Adding a new default breaks the integration tests, you'll have to add the new default field to the expected

**Comment by haircommander** [comment]
updated! thanks @JoelSpeed

**Comment by JoelSpeed** [comment]
We are asking folks to fix SSA tags as and when they come up, could you please look into fixing, for most this is just a case of adding `+listType=atomic` unless there is an obvious need for a map type
```
error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.requiredDropCapabilities must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.allowedCapabilities must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.seccompProfiles must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.fsGroup.ranges must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.groups must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.supplementalGroups.ranges must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.defaultAddCapabilities must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.volumes must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.allowedFlexVolumes must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.allowedUnsafeSysctls must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.forbiddenSysctls must set x-kubernetes-list-type
		error in security/v1/zz_generated.crd-manifests/0000_03_config-operator_01_securitycontextconstraints-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/securitycontextconstraints.security.openshift.io version/v1 field/^.users must set x-kubernetes-list-type
```

**Comment by haircommander** [comment]
updated @JoelSpeed !

**Comment by JoelSpeed** [comment]
/override ci/prow/verify-crd-schema

We can't fix the existing NoBool issues

File: security/v1/types.go
Comment (JoelSpeed) [inline security/v1/types.go:106]: Drop this line, it repeats the one above

**Comment by JoelSpeed** [comment]
LGTM once the auth team ack
/override ci/prow/verify-crd-schema

File: features/features.go
Comment (liouk) [inline features/features.go:601]: This could have gone to `authentication` as well, as it impacts PSa/SCCs; however since PSa/SCC and/or userns expertise might be required for any reported problems, I'm happy to leave this to `Node` in order to keep this coupled with the main FG `UserNamespacesSupport`.

**Review by liouk — APPROVED** [review]

**Pull request merged**
